### PR TITLE
Feature #167778283 – Keep original load date in DB

### PR DESCRIPTION
**Pull request by alfonsomunozpomer**
I reimplemented the classes `ExperimentCrud`, `ExperimentDao` and `ExperimentTrader`, and how they interact with the caches.

Now, since `scxa_experiment` has been renamed to `experiment`, I removed `ExperimentDao` and created three DAO classes repackaged by feature:
- `ExperimentCrudDao`: only concerned with CRUD operations for `ExperimentCrud`
- `PostgreSqlHealthService`: only concerned with the presence of the database
- `ExperimentTraderDao`: only concerned with getting experiment accessions by type for the trader

A single class, `ExperimentCrudDao` will be responsible for manipulatin and reading the table `experiment` and read/write to/from the updated `ExperimentDto` class.

I added `load_date` to `ExperimentDto` and `Experiment`. I also pulled `accessKey` from the DB to those two classes, so that the trader can check with the cached object if the provided access key matches. It makes the reasoning about the caching a lot simpler.

`ExperimentTrader` relies on the cached implementation of `ExperimentRepository`, whose only mission is to build experiments.

It incorporates the branch from PR https://github.com/ebi-gene-expression-group/atlas-schemas/pull/7 in the submodule `src/test/resources/schemas`.

### CODE REVIEW

File: src/test/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrudDaoIT.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomDoi;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomExperimentAccession;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomPubmedId;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomSpecies;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = TestConfig.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class ExperimentCrudDaoIT {
```
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrudDaoIT.java]: It’s something we added in a few classes as a test, and I think in general it’s a good idea. That’s why we’re including them in the API-related repos. We’ll add display names when we move the code to a dedicated service or as part of the API, don’t worry.

File: src/main/java/uk/ac/ebi/atlas/testutils/RandomDataTestUtils.java [hunk 2]
```diff
@@ -420,4 +421,12 @@ public static String generateRandomUrl() throws Exception {
 
         return uriBuilder.build().toString();
     }
+
+    public static String generateRandomPubmedId() {
```
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/testutils/RandomDataTestUtils.java]: Yes, the `testutils` package is kind of an odd one. It should only be included in tests but we’re using it from several places, and it’s grown *a lot*. In the API we currently have tests and functionality is going to be carefully added and broken down into several classes. I think there are a few methods that aren’t used any more.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrud.java [hunk 3]
```diff
@@ -1,95 +1,71 @@
...
+        return experimentCrudDao.readExperiments();
     }
 
-    public void makeExperimentPublic(String experimentAccession) {
-        setExperimentPrivacyStatus(experimentAccession, false);
+    // Update
+    @Caching(evict = {
+            @CacheEvict(cacheNames = "experiment", key = "#experimentAccession"),
+            @CacheEvict(cacheNames = "experimentAttributes", key = "#experimentAccession") })
+    public void updateExperimentPrivate(String experimentAccession, boolean isPrivate) {
```
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrud.java]: We can have more explanatory name like `updateExperimentStatus` or `updateExperimentPrivateStatus`

File: src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrudDao.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    public ImmutableList<ExperimentDto> readExperiments() {
+        LOGGER.debug("Reading all rows from table 'experiment'");
+        return ImmutableList.copyOf(
+                jdbcTemplate.query(
+                        "SELECT * FROM experiment",
+                        EXPERIMENT_DTO_ROW_MAPPER));
+    }
+
+    // Update
+    public void updateExperimentPrivate(String experimentAccession, boolean isPrivate) {
```
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrudDao.java]: Same comment as above about function name.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/baseline/BaselineExperiment.java [hunk 5]
```diff
@@ -36,11 +37,13 @@ public BaselineExperiment(@NotNull ExperimentType experimentType,
                               @NotNull List<@NotNull String> alternativeViews,
                               @NotNull List<@NotNull String> alternativeViewDescriptions,
                               @NotNull ExperimentDisplayDefaults experimentDisplayDefaults,
-                              boolean isPrivate) {
+                              boolean isPrivate,
+                              @NotNull String accessKey) {
```
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/baseline/BaselineExperiment.java]: Can we use `package-info.java` in this package as well to make code cleaner?

File: src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrud.java (hunk 3 above)
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrud.java]: I used the following naming scheme: `{Action}{Table}{Column}`. With that in mind, and knowing that `private` is a column in our experiments table, do you still find it hard to read?

File: src/main/java/uk/ac/ebi/atlas/model/experiment/baseline/BaselineExperiment.java (hunk 5 above)
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/model/experiment/baseline/BaselineExperiment.java]: Good point.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrud.java (hunk 3 above)
Comment (haideriqbal) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/ExperimentCrud.java]: It makes sense. I wasn't aware of the naming scheme.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/ExperimentDtoTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomDoi;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomExperimentAccession;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomPubmedId;
+import static uk.ac.ebi.atlas.testutils.RandomDataTestUtils.generateRandomSpecies;
+
+class ExperimentDtoTest {
+    private static final Random RNG = ThreadLocalRandom.current();
+    private static final int MAXIMUM_PUBLICATION_COUNT = 10;
+
+    static ExperimentDto generateRandomExperimentDto() {
```
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/experimentimport/ExperimentDtoTest.java]: I know this function is specific to this class but wouldn't it make more sense to move it to `RandomDataTestUtils.java` to keep consistency?

Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/ExperimentDtoTest.java]: We used to have everything in `RandomDataTestUtils.java`, which has led the class to be over 400 lines long. I think it makes it a very bit unwieldy, whereas `generateRandomExperimentDto` can be scoped nicely at the package level.

I’m experimenting a bit here and trying a slightly different approach. We’ll discuss the merits of either solution for the API project. I’d like to keep it here as an example of a random data generator that doesn’t need to be shared globally if that’s ok.

Comment (haideriqbal) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/ExperimentDtoTest.java]: I agree with your approach. It makes sense.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/idf/IdfParserIT.java [hunk 7]
```diff
@@ -16,78 +17,101 @@
...
+            bulkPopulator.addScripts(new ClassPathResource("fixtures/gxa-experiment-fixture.sql"));
+            bulkPopulator.execute(dataSource);
+        }
+
+        @AfterAll
+        void cleanDatabaseTables() {
+            ResourceDatabasePopulator populator = new ResourceDatabasePopulator();
+            populator.addScripts(new ClassPathResource("fixtures/experiment-delete.sql"));
+            populator.execute(dataSource);
+            populator.execute(dataSource);
```
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/experimentimport/idf/IdfParserIT.java]: Why are we executing the same statement twice here?

Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/idf/IdfParserIT.java]: Because I’m a sloppy developer! :smile:

**Pull request merged**